Feat/heatmap

Ollekmc opened the pull request:
Add heatMap pls review

NickKlock committed:
Add react-simple-maps modules

NickKlock committed:
Add react-simple-maps modules, Add dev Dependencies, Add Basic WorldMap.tsx, Add Routing

NickKlock committed:
Update WorldMap to add countries to user, and colorize countries by
Remove now redundant variables

Ollekmc committed:
Add heatMap w/o heat

Ollekmc committed:
Fix heatMap w/o real Users + useEffect

Ollekmc committed:
Fix heatMap real Users w/o useEffect

Ollekmc committed:
Fix heatMap & cleanCode

fannie-id commented on frontend/src/component/UserOverview.tsx:
```diff
@@ -19,12 +20,17 @@ export default function UserOverview(props:UserOverviewProps) {
             deleteUser={handleDeleteUser}
             loggedInUser={props.loggedInUser}/>
         Logged in as : {props.loggedInUser && props.loggedInUser.name}
+        <button onClick={openMap}>map </button>
```
`<button onClick={openMap}> map </button>`

NickKlock committed:
Update to implement latest main changes
Add routing for heatmap
Add navbarItem for HeatMap.tsx
Add condition for personal maps

NickKlock committed:
Change menu item name, getMethods on /traveller are now allowed for everyone

NickKlock commented on frontend/src/component/HeatMap.tsx:
```diff
@@ -0,0 +1,85 @@
+            {countryCounts.length > 0 && (
+                <Geographies geography={geoUrl}>
+                    {({geographies}) =>
+                        geographies.map((geo) => {
```
arrow functions always return a value.

NickKlock commented on frontend/src/component/HeatMap.tsx:
```diff
@@ -0,0 +1,85 @@
+    const countryList = props.countries
+    let userList:Traveller[] = useUsers().users
+
+    const [countryCounts, setCountryCounts] = useState<{
+        country: string,
+        count: number
+    }[]>([])
```
you can create a type for that

NickKlock commented on frontend/src/component/HeatMap.tsx:
```diff
@@ -0,0 +1,85 @@
+    }[]>([])
+
+    useEffect(() => {
+        const threeLetterCodeList = countryList.map(Country => Country.threeLetterCode)
+
+        function setCount(country: string) {
+            return {
+                country,
+                count: 0
+            };
+        }
+
+        let output = threeLetterCodeList.map(setCount);
+        for (const element of userList) {
+            for (const item of element.visitedCountries) {
+                const threeLetterCodeToCount = item.threeLetterCode
+                for (const element of output) if (element.country === threeLetterCodeToCount) {
+                    element.count++
+                }
+            }
+        }
+        let countVisits = []
+        for (const element of output) countVisits.push(element.count)
+        let maxVisits = Math.max(...countVisits)
+        setMaxDomain(maxVisits)
+        setCountryCounts(output)
+    }, [])
```
Someone else then your probably needs a day to understand your code here. maybe less for-each and more filter(), map() etc?

NickKlock commented on frontend/src/component/HeatMap.tsx:
```diff
@@ -0,0 +1,85 @@
+            };
+        }
+
+        let output = threeLetterCodeList.map(setCount);
```
what is output suppose to mean? maybe u can be a bit more creative with your variable names.
